src/amalgam/gen/neondot.c

```diff
@@ -5958,3 +5962,160 @@ void xnn_qu8_igemm_minmax_rndnu_ukernel_4x8c4__neondot(
+  const float scalar_max = params->scalar.max;
+
+  if (m == 0) {
+      return;
+  }
+
+  const size_t num_rows = m;
+  const size_t num_cols = n;
+
+  const size_t lhs_packed_stride = kai_mr * (k + sizeof(float) + sizeof(float));
+
+  const int8x16_t nibble_mask = vdupq_n_s8(0xF0);
+
+  const uint8_t* lhs_ptr_start = lhs_p;
+
+  for (size_t row_idx = 0; row_idx < num_rows; row_idx += kai_mr) {
+      const uint8_t* rhs_ptr = rhs_p;
+      for (size_t col_idx = 0; col_idx < num_cols; col_idx += kai_nr) {
+
+          const uint8_t* lhs_ptr = lhs_ptr_start;
+
+          // Main f32 accumulator
+          int32x4_t iacc0011 = vdupq_n_s32(0);
+          int32x4_t iacc2233 = vdupq_n_s32(0);
+          int32x4_t iacc4455 = vdupq_n_s32(0);
+          int32x4_t iacc6677 = vdupq_n_s32(0);
+
+          for (size_t b = 0; b < k; b += kai_k0) {
+              // Set up RHS
+              const int8x16_t rhs_raw_vec_0 = vld1q_s8((const int8_t*)(rhs_ptr + 0));
+              const int8x16_t rhs_raw_vec_1 = vld1q_s8((const int8_t*)(rhs_ptr + 16));
+              const int8x16_t rhs_raw_vec_2 = vld1q_s8((const int8_t*)(rhs_ptr + 32));
+              const int8x16_t rhs_raw_vec_3 = vld1q_s8((const int8_t*)(rhs_ptr + 48));
+              const int8x16_t rhs_raw_vec_4 = vld1q_s8((const int8_t*)(rhs_ptr + 64));
+              const int8x16_t rhs_raw_vec_5 = vld1q_s8((const int8_t*)(rhs_ptr + 80));
+              const int8x16_t rhs_raw_vec_6 = vld1q_s8((const int8_t*)(rhs_ptr + 96));
+              const int8x16_t rhs_raw_vec_7 = vld1q_s8((const int8_t*)(rhs_ptr + 112));
+
+              // Low nibble
+              const int8x16_t rhs_vec_0_0 = vshlq_n_s8(rhs_raw_vec_0, 4);
+              const int8x16_t rhs_vec_1_0 = vshlq_n_s8(rhs_raw_vec_1, 4);
+              const int8x16_t rhs_vec_2_0 = vshlq_n_s8(rhs_raw_vec_2, 4);
+              const int8x16_t rhs_vec_3_0 = vshlq_n_s8(rhs_raw_vec_3, 4);
+              const int8x16_t rhs_vec_4_0 = vshlq_n_s8(rhs_raw_vec_4, 4);
+              const int8x16_t rhs_vec_5_0 = vshlq_n_s8(rhs_raw_vec_5, 4);
+              const int8x16_t rhs_vec_6_0 = vshlq_n_s8(rhs_raw_vec_6, 4);
+              const int8x16_t rhs_vec_7_0 = vshlq_n_s8(rhs_raw_vec_7, 4);
+
+              // High nibble
+              const int8x16_t rhs_vec_0_1 = vandq_s8(rhs_raw_vec_0, nibble_mask);
+              const int8x16_t rhs_vec_1_1 = vandq_s8(rhs_raw_vec_1, nibble_mask);
+              const int8x16_t rhs_vec_2_1 = vandq_s8(rhs_raw_vec_2, nibble_mask);
+              const int8x16_t rhs_vec_3_1 = vandq_s8(rhs_raw_vec_3, nibble_mask);
+              const int8x16_t rhs_vec_4_1 = vandq_s8(rhs_raw_vec_4, nibble_mask);
+              const int8x16_t rhs_vec_5_1 = vandq_s8(rhs_raw_vec_5, nibble_mask);
+              const int8x16_t rhs_vec_6_1 = vandq_s8(rhs_raw_vec_6, nibble_mask);
+              const int8x16_t rhs_vec_7_1 = vandq_s8(rhs_raw_vec_7, nibble_mask);
+
+              const int8x16_t lhs_vec_0 = vld1q_s8((const int8_t*)(lhs_ptr + 0));
```

Comment:
This can do the transpose using ld2 lane, like the current i8mm gemm, avoiding the combine below.